keel-docker/src/main/kotlin/com/netflix/spinnaker/keel/docker/DockerImageResolver.kt

```diff
@@ -53,37 +55,47 @@ abstract class DockerImageResolver<T : ResourceSpec>(
   /**
    * Replace exiting container with resolved container
    */
-  abstract fun updateContainerInSpec(resource: Resource<T>, container: Container): Resource<T>
+  abstract fun updateContainerInSpec(resource: Resource<T>, container: ContainerProvider): Resource<T>
 
   /**
    * Get all available the tags for an image
    */
   abstract fun getTags(account: String, organization: String, image: String): List<String>
 
-  /**
+  /**De
    * Get the digest for a specific image
    */
   abstract fun getDigest(account: String, organization: String, image: String, tag: String): String
 
   override fun invoke(resource: Resource<T>): Resource<T> {
-    var container = getContainerFromSpec(resource)
-    if (container is ContainerWithDigest) {
+    val container = getContainerFromSpec(resource)
+    if (container is DigestProvider) {
       return resource
     }
-    container = container as ContainerWithVersionedTag
-    val account = getAccountFromSpec(resource)
     val deliveryConfig = deliveryConfigRepository.deliveryConfigFor(resource.id)
     val environment = deliveryConfigRepository.environmentFor(resource.id)
-    val tag: String = if (deliveryConfig != null && environment != null) {
-      findTagGivenDeliveryConfig(deliveryConfig, environment, container.toArtifact())
-    } else {
-      findTagGivenStrategy(account, container)
-    }
+    val artifact = getArtifact(container, deliveryConfig)
+    val account = getAccountFromSpec(resource)
+    val tag: String = findTagGivenDeliveryConfig(deliveryConfig, environment, artifact)
 
-    val newContainer = getContainer(account, container, tag)
+    val newContainer = getContainer(account, artifact, tag)
     return updateContainerInSpec(resource, newContainer)
   }
 
+  fun getArtifact(container: ContainerProvider, deliveryConfig: DeliveryConfig): DockerArtifact =
+    when (container) {
+      is ReferenceProvider -> {
```

Comment:
This is the bit that looks up the actual artifact based on a reference in your `container` section of your cluster spec